release/version.py

```diff
@@ -0,0 +1,99 @@
+#!/usr/bin/env python3
+
+import os
+import sys
+from typing import Optional
+import re
+import subprocess
+
+GIT_VERSION_FILE = "GIT_VERSION"
+
+
+def git() -> str:
+    """get the version from git. Can be
+    - X.Y.Z if a tag is set on the current HEAD
+    - X.Y.Z-<count>-g<commitid> otherwise where
+        - X.Y.Z is the latest version tag found
+        - <count> is the number of commits since then
+        - <commitid> is the current HEAD commitid
+    """
+    cmd = "git describe --tags --match=[0-9]*.[0-9]*.[0-9]*"
+    try:
+        p = subprocess.run(cmd.split(" "), capture_output=True, check=True)
+        v = p.stdout.decode("utf-8").strip()
+        # tox is running setup.py sdist from the git repo, and then runs again outside
+        # of the git repo. At this second step, we cannot run git commands.
+        # So we save the git version in a file and include it in the source distribution
+        with open(GIT_VERSION_FILE, "w") as f:
+            f.write(v)
+        return v
+    except subprocess.CalledProcessError as e:
```

Comment:
Not 100% sure if I am right here, but given the comment regarding calling tox twice, this looks like you are using exceptions for flow control. I would prefer having the read operation outside the except block since you can clearly specify the conditions for when it should be called.

Reply:
not really: i always want to prefer the output of `git describe` than reading the file. How would you specify the conditions (i thought of subprocessing `git rev-parse`, but in the end it's the same than here..
The alternative could be `check=False` and check the `errorcode`. Is that what you think about ?

Reply:
or ?
```python
    except subprocess.CalledProcessError as e:
        if not os.path.exists(GIT_VERSION_FILE):
            print(e.stderr)
            raise e
    with open(GIT_VERSION_FILE, "r") as f:
        return f.read()
```

Reply:
Hm, perhaps check if it's a git repository and if not try reading the file? Bash example for checking if dir is a git repo:  
```
if [ -d .git ]; then
  echo .git;
else
  git rev-parse --git-dir 2> /dev/null;
fi;
```

Reply:
so in the end we'd run a subcommand and check the outcome anyways. Is it really better ?